Sources/NetworkProtection/Diagnostics/KeyExpirationTester.swift

```
@@ -0,0 +1,164 @@
+        await scheduleTimer(testImmediately: testImmediately)
+    }
+
+    func stop() {
+        Logger.networkProtectionConnectionTester.log("🔴 Stopping rekey timer")
+        stopScheduledTimer()
+        isRunning = false
+    }
+
+    // MARK: - Timer scheduling
+
+    private func scheduleTimer(testImmediately: Bool) async {
+        stopScheduledTimer()
+
+        if testImmediately {
+            await rekeyIfExpired()
+        }
+
+        task = Task.periodic(interval: intervalBetweenTests) { [weak self] in
+            await self?.rekeyIfExpired()
+        }
+    }
+
+    private func stopScheduledTimer() {
+        task?.cancel()
+        task = nil
+    }
+
+    // MARK: - Testing the connection
+
+    private var isKeyExpired: Bool {
+        guard let currentExpirationDate = keyStore.currentExpirationDate else {
+            return true
+        }
+
+        return currentExpirationDate <= Date()
+    }
+
+    // MARK: - Expiration check
+
+    func rekeyIfExpired() async {
+
+        guard !isTestingExpiration else {
+            return
+        }
+
+        isTestingExpiration = true
+
+        defer {
+            isTestingExpiration = false
+        }
+
+        guard await canRekey() else {
+            Logger.networkProtectionKeyManagement.log("Can't rekey right now as some preconditions aren't met.")
+            return
+        }
+
+        Logger.networkProtectionKeyManagement.log("Checking if rekey is necessary...")
+
+        guard isKeyExpired else {
+            Logger.networkProtectionKeyManagement.log("The key is not expired")
+            return
+        }
+
+        Logger.networkProtectionKeyManagement.log("Rekeying now.")
+        do {
+            try await rekey()
+            Logger.networkProtectionKeyManagement.log("Rekeying completed.")
+        } catch {
+            Logger.networkProtectionKeyManagement.error("Rekeying failed with error: \(error, privacy: .public).")
+        }
```

Should we surface any errors outside of the tester so that the caller can fire a pixel or take some other action?

---

This is a good observation.

I've thought about this but it's unclear to me right now that this is needed, considering rekey failures are already recorded in the `rekey()` call.  Open to discussing more.